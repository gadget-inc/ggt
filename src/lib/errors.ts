import type { Config } from "@oclif/core";
import type { OptionFlag } from "@oclif/core/lib/interfaces";
import cleanStack from "clean-stack";
import { HTTPError } from "got/dist/source";
import type { GraphQLError } from "graphql";
import { isArray, isBoolean, isError, isNil, isNumber, isString, uniqBy } from "lodash";
import newGithubIssueUrl from "new-github-issue-url";
import { serializeError as baseSerializeError } from "serialize-error";
import dedent from "ts-dedent";
import type { SetOptional, Writable } from "type-fest";
import { inspect } from "util";
import type { CloseEvent, ErrorEvent } from "ws";
import type Sync from "../commands/sync";
import type { Payload } from "./client";

/**
 * Base class for all errors.
 *
 * Inspired by gadget's GadgetError and oclif's PrettyPrintableError.
 */
export abstract class BaseError extends Error {
  /**
   * A GGT_CLI_SOMETHING human/machine readable unique identifier for this error.
   */
  code: string;

  /**
   * The underlying *thing* that caused this error.
   */
  override cause?: any;

  /**
   * Assume the stack trace exists.
   */
  override stack!: string;

  /**
   * Indicates whether this error is considered a bug or not.
   */
  abstract isBug: IsBug;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    Error.captureStackTrace(this, this.constructor);
  }

  /**
   * Turns this error into a user-friendly message that explains what went wrong and how to fix it. A good write up of what an error should
   * look like can be found here: {@link https://clig.dev/#errors}
   */
  render(config: Config): string {
    const header = this.header(config);
    const body = this.body(config);
    const footer = this.footer(config);

    let output = "";
    if (header) output += header + "\n\n";
    if (body) output += body + "\n\n";
    if (footer) output += footer + "\n\n";

    return output;
  }

  protected header(_: Config): string {
    return `${this.code}: ${this.message}`;
  }

  protected footer(config: Config): string {
    switch (this.isBug) {
      case IsBug.NO:
        return "";
      case IsBug.MAYBE:
        return dedent`
          If you think this is a bug, please submit an issue using the link below.

          ${newGithubIssueUrl(this.issueOptions(config))}
        `;
      case IsBug.YES:
        return dedent`
          This is a bug :(

          Visit the link below to see if someone has already reported this issue.
          https://github.com/gadget-inc/ggt/issues?q=is%3Aissue+is%3Aopen+label%3Abug+${this.code}

          If nobody has, you can submit one using the link below.
          ---
          ${newGithubIssueUrl(this.issueOptions(config))}
        `;
    }
  }

  protected issueOptions(config: Config): newGithubIssueUrl.Options {
    const options = {
      repoUrl: "https://github.com/gadget-inc/ggt",
      title: `[BUG ${this.code}]: `,
      labels: ["bug"],
      body: dedent`
        ### Description
        [Please describe what you were doing when this error occurred]

        ### Command
        \`\`\`sh-session
        ${config.bin} ${process.argv.slice(2).join(" ")}
        \`\`\`

        ### Environment
        \`\`\`
        version: ${config.version}
        platform: ${config.platform}
        arch: ${config.arch}
        shell: ${config.shell}
        timestamp: ${new Date().toISOString()}
        \`\`\`

        ### Stack Trace
        \`\`\`
        ${cleanStack(this.stack)}
        \`\`\`
      `,
    };

    if (this.cause) {
      options.body += dedent`

        ### Cause
        \`\`\`json
        ${JSON.stringify(isError(this.cause) ? serializeError(this.cause) : this.cause, null, 2)}
        \`\`\`
      `;
    }

    return options;
  }

  protected abstract body(_: Config): string;
}

/**
 * Universal Error object to json blob serializer.
 * Wraps `serialize-error` with some handy stuff, like special support for Got HTTP errors
 */
export function serializeError(error: Error | string | unknown): Record<string, any> {
  let serialized = baseSerializeError(error);
  if (typeof serialized == "string") {
    serialized = { message: serialized };
  }

  if (error instanceof HTTPError && error.name === "RequestError") {
    delete serialized["timings"];
    serialized["options"] = { method: error.options.method, url: error.options.url.toJSON() };
    serialized["responseBody"] = inspect(error.response?.body);
  }

  return serialized;
}

export enum IsBug {
  NO,
  MAYBE,
  YES,
}

/**
 * Our "catch all" error. If this error is thrown, we almost certainly have a bug.
 *
 * Whenever possible, we should use a more specific error so that we can provide more useful information.
 */
export class UnexpectedError extends BaseError {
  isBug = IsBug.YES;

  constructor(override cause: Error) {
    super("GGT_CLI_UNEXPECTED_ERROR", "An unexpected error occurred");
  }

  protected body(_: Config): string {
    return cleanStack(this.cause.stack ?? this.stack);
  }
}

export class ClientError extends BaseError {
  isBug = IsBug.MAYBE;

  constructor(readonly payload: Payload<any, any>, override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent) {
    super("GGT_CLI_CLIENT_ERROR", "An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct them into an object. We discard the `target` property because it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as any;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as any;
    }
  }

  override body(_: Config): string {
    if (isGraphQLErrors(this.cause)) {
      if (this.cause.length > 1) {
        let output = "Gadget responded with multiple errors:";
        for (const error of uniqBy(this.cause, "message")) {
          output += `\n * ${error.message}`;
        }
        return output;
      } else {
        return dedent`
          Gadget responded with an unexpected error.

          ${this.cause[0]?.message}
        `;
      }
    }

    if (isCloseEvent(this.cause)) {
      return "The connection to Gadget closed unexpectedly.";
    }

    if (isErrorEvent(this.cause)) {
      return dedent`
          The connection to Gadget received an unexpected error.

          ${this.cause.message}
      `;
    }

    if (isError(this.cause)) {
      return dedent`
          An unexpected error occurred.

          ${this.cause.message}
      `;
    }

    return this.cause;
  }

  protected override issueOptions(config: Config): newGithubIssueUrl.Options {
    const options = super.issueOptions(config) as Writable<newGithubIssueUrl.Options>;
    options.body += dedent`

      ### GraphQL

      #### Query
      \`\`\`graphql
      ${this.payload.query}
      \`\`\`
    `;

    // mutations can have large/sensitive payloads, so we don't include them by default
    if (!this.payload.query.trimStart().startsWith("mutation")) {
      options.body += dedent`

        #### Variables
        \`\`\`json
        ${JSON.stringify(this.payload.variables, null, 2)}
        \`\`\`
        `;
    }

    return options;
  }
}

export class YarnNotFoundError extends BaseError {
  isBug = IsBug.NO;

  constructor() {
    super("GGT_CLI_YARN_NOT_FOUND", "Yarn not found");
  }

  protected body(_: Config): string {
    return dedent`
      Yarn is required to sync your application.

      Please install Yarn by running:

        $ npm install --global yarn

      For more information, see: https://classic.yarnpkg.com/en/docs/install
    `;
  }
}

export class FlagError extends BaseError {
  isBug = IsBug.NO;

  constructor(readonly flag: Partial<OptionFlag<unknown>>, readonly description: string) {
    const name = flag.char ? `-${flag.char}, --${flag.name}` : `--${flag.name}`;
    super("GGT_CLI_FLAG_ERROR", `Invalid value provided for the ${name} flag`);
  }

  protected body(_: Config): string {
    return this.description;
  }
}

export class InvalidSyncFileError extends BaseError {
  isBug = IsBug.MAYBE;

  constructor(override readonly cause: unknown, readonly sync: Sync, readonly app: string | undefined) {
    super("GGT_CLI_INVALID_SYNC_FILE", "The .ggt/sync.json file was invalid or not found");
  }

  protected body(_: Config): string {
    return dedent`
      We failed to read the Gadget metadata file in this directory:

        ${this.sync.dir}

      If you're running \`ggt sync\` for the first time, we recommend using an empty directory such as:

        ~/gadget/${this.app || "<name of app>"}

      Otherwise, if you're sure you want to sync the contents of "${
        this.sync.dir
      }" to Gadget, run \`ggt sync\` again with the \`--force\` flag:

        $ ggt sync ${this.sync.argv.join(" ")} --force

      You will be prompted to either merge your local files with your remote ones or reset your local files to your remote ones.
    `;
  }
}

function isCloseEvent(e: any): e is SetOptional<CloseEvent, "target"> {
  return !isNil(e) && isString(e.type) && isNumber(e.code) && isString(e.reason) && isBoolean(e.wasClean);
}

function isErrorEvent(e: any): e is SetOptional<ErrorEvent, "target"> {
  return !isNil(e) && isString(e.type) && isString(e.message) && isError(e.error);
}

function isGraphQLErrors(e: any): e is readonly GraphQLError[] {
  return isArray(e) && e.every((e) => !isNil(e) && isString(e.message) && isArray(e.locations) && isArray(e.path));
}
