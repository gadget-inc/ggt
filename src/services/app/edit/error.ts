import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { sprint } from "../../output/print.js";
import { CLIError, IsBug } from "../../output/report.js";
import { uniq } from "../../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString } from "../../util/is.js";
import { serializeError } from "../../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./operation.js";

export class EditError extends CLIError {
  isBug = IsBug.MAYBE;

  override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
  ) {
    super("An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body = "";

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}
