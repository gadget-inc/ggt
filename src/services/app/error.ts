import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { GGTError, IsBug } from "../output/report.js";
import { sprint } from "../output/sprint.js";
import { uniq } from "../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString, isStringArray } from "../util/is.js";
import { serializeError } from "../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./edit/operation.js";

export class ClientError extends GGTError {
  public isBug: IsBug;

  override cause: string | string[] | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
    isBug?: IsBug,
  ) {
    super("An error occurred while communicating with Gadget");

    if (isBug) {
      this.isBug = isBug;
    } else {
      this.isBug = IsBug.MAYBE;
    }

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isStringArray(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body: string;

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      case isStringArray(this.cause):
        if (this.cause.length === 1) {
          body = String(this.cause[0]);
        } else {
          body = this.cause.join(", ");
        }
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}

export class AuthenticationError extends ClientError {
  constructor(request: GraphQLQuery | GraphQLMutation | GraphQLSubscription) {
    super(request, "Request authentication failed due to the session expiring while running the command. Please sign-in again.", IsBug.NO);
  }
}
