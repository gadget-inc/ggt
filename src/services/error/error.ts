import arg from "arg";
import cleanStack from "clean-stack";
import type { GraphQLError } from "graphql";
import assert from "node:assert";
import { randomUUID } from "node:crypto";
import { dedent } from "ts-dedent";
import type { JsonObject } from "type-fest";
import type { CloseEvent, ErrorEvent } from "ws";
import type { Query } from "../app/edit-graphql.js";
import { env } from "../config/env.js";
import { sprintln } from "../output/sprint.js";
import { compact, uniq } from "../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString } from "../util/is.js";
import { serializeError } from "../util/object.js";

export enum IsBug {
  YES = "yes",
  NO = "no",
  MAYBE = "maybe",
}

/**
 * Base class for all errors.
 */
export abstract class CLIError extends Error {
  /**
   * A GGT_CLI_SOMETHING human/machine readable unique identifier for this error.
   */
  code: string;

  /**
   * The Sentry event ID for this error.
   */
  sentryEventId = env.testLike ? "00000000-0000-0000-0000-000000000000" : randomUUID();

  /**
   * The underlying *thing* that caused this error.
   */
  cause?: unknown;

  /**
   * Assume the stack trace exists.
   */
  override stack!: string;

  /**
   * Indicates whether this error is considered a bug or not.
   */
  abstract isBug: IsBug;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    Error.captureStackTrace(this, this.constructor);
  }

  /**
   * Constructs a CLIError from a cause.
   */
  static from(cause: unknown): CLIError {
    if (cause instanceof CLIError) {
      return cause;
    }
    if (cause instanceof arg.ArgError) {
      return new ArgError(cause.message);
    }
    return new UnexpectedError(cause);
  }

  /**
   * Turns this error into a user-friendly message that explains what
   * went wrong and how to fix it. A good write up of what an error
   * should look like can be found here:
   * {@link https://clig.dev/#errors}
   */
  render(): string {
    return compact([this.header(), this.body(), this.footer()]).join("\n\n");
  }

  protected header(): string {
    return `${this.code}: ${this.message}`;
  }

  protected footer(): string {
    if (this.isBug === IsBug.NO) {
      return "";
    }

    return dedent`
      ${this.isBug === IsBug.YES ? "This is a bug" : "If you think this is a bug"}, please submit an issue using the link below.

      https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=${this.sentryEventId}
    `;
  }

  protected abstract body(): string;
}

/**
 * Our "catch all" error.
 *
 * If this error is thrown, we almost certainly have a bug, and should
 * either fix it or add a more specific error so that we can provide
 * more useful information.
 */
export class UnexpectedError extends CLIError {
  isBug = IsBug.YES;

  constructor(override cause: unknown) {
    super("GGT_CLI_UNEXPECTED_ERROR", "An unexpected error occurred");
  }

  protected body(): string {
    if (isError(this.cause)) {
      return cleanStack(this.cause.stack ?? this.stack);
    }
    return this.stack;
  }
}

export class EditGraphQLError extends CLIError {
  isBug = IsBug.MAYBE;

  override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly query: Query<JsonObject>,
    cause: unknown,
  ) {
    super("GGT_CLI_CLIENT_ERROR", "An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override body(): string {
    if (isGraphQLErrors(this.cause)) {
      const errors = uniq(this.cause.map((x) => x.message));
      if (errors.length > 1) {
        let n = 1;
        return sprintln("Gadget responded with multiple errors:").concat(`\n  ${n++}. ${errors.join(`\n  ${n++}. `)}`);
      } else {
        return dedent`
          Gadget responded with the following error:

            ${errors[0]}
        `;
      }
    }

    if (isCloseEvent(this.cause)) {
      return "The connection to Gadget closed unexpectedly.";
    }

    if (isErrorEvent(this.cause) || isError(this.cause)) {
      return this.cause.message;
    }

    return this.cause;
  }
}

export class YarnNotFoundError extends CLIError {
  isBug = IsBug.NO;

  constructor() {
    super("GGT_CLI_YARN_NOT_FOUND", "Yarn not found");
  }

  protected body(): string {
    return dedent`
      Yarn must be installed to sync your application. You can install it by running:

        $ npm install --global yarn

      For more information, see: https://classic.yarnpkg.com/en/docs/install
    `;
  }
}

export class ArgError extends CLIError {
  isBug = IsBug.NO;

  constructor(message: string) {
    super("GGT_CLI_ARG_ERROR", message);
  }

  // eslint-disable-next-line lodash/prefer-constant
  protected override header(): string {
    return "";
  }

  protected override body(): string {
    return this.message;
  }
}

export class InvalidSyncFileError extends CLIError {
  isBug = IsBug.MAYBE;

  constructor(
    readonly dir: string,
    readonly app: string | undefined,
  ) {
    super("GGT_CLI_INVALID_SYNC_FILE", "The .gadget/sync.json file was invalid or not found");
    this.app ??= "<name of app>";
  }

  protected body(): string {
    return dedent`
      We failed to read the Gadget metadata file in this directory:

        ${this.dir}

      If you're running \`ggt sync\` for the first time, we recommend using an empty directory such as:

        ~/gadget/${this.app}

      Otherwise, if you're sure you want to sync the contents of that directory to Gadget, run \`ggt sync\` again with the \`--force\` flag:

        $ ggt sync --app ${this.app} ${this.dir} --force

      You will be prompted to either merge your local files with your remote ones or reset your local files to your remote ones.
    `;
  }
}
