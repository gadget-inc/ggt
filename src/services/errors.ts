import * as Sentry from "@sentry/node";
import arg from "arg";
import cleanStack from "clean-stack";
import { RequestError } from "got";
import type { GraphQLError } from "graphql";
import { compact, uniqBy } from "lodash";
import { randomUUID } from "node:crypto";
import os from "node:os";
import { inspect, isArray, isError } from "node:util";
import { serializeError as baseSerializeError, type ErrorObject } from "serialize-error";
import { dedent } from "ts-dedent";
import type { JsonObject, SetOptional } from "type-fest";
import type { CloseEvent, ErrorEvent } from "ws";
import { z } from "zod";
import type { App } from "./app.js";
import { config, env } from "./config.js";
import type { Payload } from "./edit-graphql.js";
import type { User } from "./user.js";

let app: App | undefined;
let user: User | undefined;

export const setUser = (newUser: User | undefined): void => {
  user = newUser;
  // eslint-disable-next-line unicorn/no-null
  Sentry.setUser(newUser ?? null);
};

export const setApp = (newApp: App | undefined): void => {
  app = newApp;
};

/**
 * Base class for all errors.
 */
export abstract class CLIError extends Error {
  /**
   * A GGT_CLI_SOMETHING human/machine readable unique identifier for this error.
   */
  code: string;

  /**
   * The Sentry event ID for this error.
   */
  sentryEventId = env.testLike ? "00000000-0000-0000-0000-000000000000" : randomUUID();

  /**
   * The underlying *thing* that caused this error.
   */
  cause?: unknown;

  /**
   * Assume the stack trace exists.
   */
  override stack!: string;

  /**
   * Indicates whether this error is considered a bug or not.
   */
  abstract isBug: IsBug;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    Error.captureStackTrace(this, this.constructor);
  }

  /**
   * Constructs a CLIError from a cause.
   */
  static from(cause: unknown): CLIError {
    if (cause instanceof CLIError) return cause;
    if (cause instanceof arg.ArgError) return new ArgError(cause.message);
    return new UnexpectedError(cause);
  }

  async capture(): Promise<void> {
    if (this.isBug == IsBug.NO) return;

    Sentry.getCurrentHub().captureException(this, {
      event_id: this.sentryEventId,
      captureContext: {
        user: user ? { id: String(user.id), email: user.email, username: user.name ?? undefined } : undefined,
        tags: {
          applicationId: app ? app.id : undefined,
          arch: config.arch,
          isBug: this.isBug,
          code: this.code,
          environment: env.value,
          platform: config.platform,
          shell: config.shell,
          version: config.version,
        },
        contexts: {
          cause: this.cause ? serializeError(this.cause) : undefined,
          app: {
            command: `ggt ${process.argv.slice(2).join(" ")}`,
            argv: process.argv,
          },
          device: {
            name: os.hostname(),
            family: os.type(),
            arch: os.arch(),
          },
          runtime: {
            name: process.release.name,
            version: process.version,
          },
        },
      },
    });

    await Sentry.flush(2000);
  }

  /**
   * Turns this error into a user-friendly message that explains what went wrong and how to fix it. A good write up of
   * what an error should look like can be found here: {@link https://clig.dev/#errors}
   */
  render(): string {
    return compact([this.header(), this.body(), this.footer()]).join("\n\n");
  }

  protected header(): string {
    return `${this.code}: ${this.message}`;
  }

  protected footer(): string {
    if (this.isBug == IsBug.NO) return "";

    return dedent`
      ${this.isBug == IsBug.YES ? "This is a bug" : "If you think this is a bug"}, please submit an issue using the link below.

      https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=${this.sentryEventId}
    `;
  }

  protected abstract body(): string;
}

/**
 * Universal Error object to json blob serializer.
 *
 * Wraps `serialize-error` with some handy stuff, like special support
 * for Got HTTP errors
 */
export function serializeError(error: unknown): ErrorObject {
  let serialized = baseSerializeError(isArray(error) ? new AggregateError(error) : error);
  if (typeof serialized == "string") {
    serialized = { message: serialized };
  }

  if (error instanceof RequestError) {
    serialized["timings"] = undefined;
    serialized["options"] = {
      method: error.options.method,
      url: error.options.url instanceof URL ? error.options.url.toJSON() : error.options.url,
    };
    serialized["responseBody"] = inspect(error.response?.body);
  }

  return serialized;
}

export enum IsBug {
  YES = "yes",
  NO = "no",
  MAYBE = "maybe",
}

/**
 * Our "catch all" error. If this error is thrown, we almost certainly have a bug.
 *
 * Whenever possible, we should use a more specific error so that we can provide more useful information.
 */
export class UnexpectedError extends CLIError {
  isBug = IsBug.YES;

  constructor(override cause: unknown) {
    super("GGT_CLI_UNEXPECTED_ERROR", "An unexpected error occurred");
  }

  protected body(): string {
    if (isError(this.cause)) {
      return cleanStack(this.cause.stack ?? this.stack);
    }
    return this.stack;
  }
}

export class ClientError extends CLIError {
  isBug = IsBug.MAYBE;

  constructor(
    readonly payload: Payload<JsonObject, JsonObject>,
    override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent,
  ) {
    super("GGT_CLI_CLIENT_ERROR", "An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    }
  }

  override body(): string {
    if (isGraphQLErrors(this.cause)) {
      if (this.cause.length > 1) {
        const errors = uniqBy(this.cause, "message");

        let output = "Gadget responded with multiple errors:\n";
        for (let i = 0; i < errors.length; i++) {
          output += `\n  ${i + 1}. ${errors[i]?.message}`;
        }

        return output;
      } else {
        return dedent`
          Gadget responded with the following error:

            ${this.cause[0]?.message}
        `;
      }
    }

    if (isCloseEvent(this.cause)) {
      return "The connection to Gadget closed unexpectedly.";
    }

    if (isErrorEvent(this.cause) || isError(this.cause)) {
      return this.cause.message;
    }

    return this.cause;
  }
}

export class YarnNotFoundError extends CLIError {
  isBug = IsBug.NO;

  constructor() {
    super("GGT_CLI_YARN_NOT_FOUND", "Yarn not found");
  }

  protected body(): string {
    return dedent`
      Yarn must be installed to sync your application. You can install it by running:

        $ npm install --global yarn

      For more information, see: https://classic.yarnpkg.com/en/docs/install
    `;
  }
}

export class ArgError extends CLIError {
  isBug = IsBug.NO;

  constructor(message: string) {
    super("GGT_CLI_ARG_ERROR", message);
  }

  // eslint-disable-next-line lodash/prefer-constant
  protected override header(): string {
    return "";
  }

  protected override body(): string {
    return this.message;
  }
}

export class InvalidSyncFileError extends CLIError {
  isBug = IsBug.MAYBE;

  constructor(
    readonly dir: string,
    readonly app: string | undefined,
  ) {
    super("GGT_CLI_INVALID_SYNC_FILE", "The .gadget/sync.json file was invalid or not found");
    this.app ??= "<name of app>";
  }

  protected body(): string {
    return dedent`
      We failed to read the Gadget metadata file in this directory:

        ${this.dir}

      If you're running \`ggt sync\` for the first time, we recommend using an empty directory such as:

        ~/gadget/${this.app}

      Otherwise, if you're sure you want to sync the contents of that directory to Gadget, run \`ggt sync\` again with the \`--force\` flag:

        $ ggt sync --app ${this.app} ${this.dir} --force

      You will be prompted to either merge your local files with your remote ones or reset your local files to your remote ones.
    `;
  }
}

function isCloseEvent(e: unknown): e is SetOptional<CloseEvent, "target"> {
  return z.object({ type: z.string(), code: z.number(), reason: z.string(), wasClean: z.boolean() }).safeParse(e).success;
}

function isErrorEvent(e: unknown): e is SetOptional<ErrorEvent, "target"> {
  return z.object({ type: z.string(), message: z.string(), error: z.any() }).safeParse(e).success;
}

function isGraphQLErrors(e: unknown): e is readonly GraphQLError[] {
  return z.array(z.object({ message: z.string() })).safeParse(e).success;
}
