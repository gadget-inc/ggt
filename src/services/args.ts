import _ from "lodash";
import { ArgError } from "./errors.js";
import { sprint } from "./output.js";

export const parseBoolean = (value: string) => _.includes(["true", "1"], _.toLower(_.trim(value)));

export const App = (value: string, name: string) => {
  const slug = /^(https:\/\/)?(?<slug>[\w-]+?)(--development)?(\..*)?$/.exec(value)?.groups?.["slug"];
  if (slug) {
    return slug;
  }

  throw new ArgError(
    sprint`
      The ${name} option must be the application's slug or URL

      Examples:

        --app my-app
        --app my-app.gadget.app
        --app https://my-app.gadget.app
        --app https://my-app.gadget.app/edit
        --app https://my-app--development.gadget.app/edit
    `,
  );
};
