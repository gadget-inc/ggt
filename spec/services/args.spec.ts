import { describe, expect, it } from "vitest";
import { AppArg, parseBoolean } from "../../src/services/args.js";
import { ArgError } from "../../src/services/errors.js";
import { expectError } from "../__support__/error.js";

describe("args", () => {
  describe("parseBoolean", () => {
    it.each(["1", "true", "True", "TRUE"])("parses %s as true", (value) => {
      expect(parseBoolean(value)).toBe(true);
    });

    it.each(["0", "false", "False", "FALSE"])("parses %s as false", (value) => {
      expect(parseBoolean(value)).toBe(false);
    });
  });

  describe("AppArg", () => {
    it.each([
      "my-app",
      "my-app.gadget.app",
      "my-app--development.gadget.app",
      "https://my-app.gadget.app",
      "https://my-app--development.gadget.app",
      "https://my-app.gadget.app/edit",
      "https://my-app.gadget.app/edit/files/routes/GET.js",
    ])("accepts %s", (value) => {
      expect(AppArg(value, "app")).toEqual("my-app");
    });

    it.each(["~"])("rejects %s", async (value) => {
      const error = await expectError(() => AppArg(value, "app"));
      expect(error).toBeInstanceOf(ArgError);
      expect(error.message).toMatchSnapshot();
    });
  });
});
