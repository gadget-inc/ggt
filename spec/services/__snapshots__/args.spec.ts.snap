// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`args > AppArg > rejects ~ 1`] = `
"The app option must be the application's slug or URL

Examples:

  --app my-app
  --app my-app.gadget.app
  --app https://my-app.gadget.app
  --app https://my-app.gadget.app/edit
  --app https://my-app--development.gadget.app/edit"
`;
